scheduler/src/cook/hooks.clj

```diff
@@ -0,0 +1,174 @@
+    ; invoked and it sinks or swims. Short circuits the evaluation, done below.
+    (or is-aged-out?
+        ;; Ok. Not aging out. Query the underlying plugin as to the status.
+        (let [{:keys [status] :as raw-result} (check-job-invocation hook-object job)
+              result
+              (if (or not-found?
+                      (= status :accepted))
+                ; If not found, or we got an accepted status, store it and reset the counters.
+                (merge raw-result
+                       {:last-seen (t/now)
+                        :first-seen (t/now)
+                        :seen-count 1})
+                ; We were found, but didn't get an accepted status. Increment the counters for aging out.
+                (merge raw-result
+                       {:last-seen (t/now)
+                        :first-seen first-seen
+                        :seen-count (inc seen-count)}))]
+          (assert (#{:accepted :deferred} status) (str "Plugin must return a status of :accepted or :deferred. Got " status))
+
+          (ccache/put-cache! job-invocations-cache :job/uuid job
+                             result)
+          ; Did the query, If the status was accepted, then we're keeping it.
+          (= status :accepted)))))
+
+(defn filter-job-invocations
+  "Run the hooks for a set of jobs at invocation time, returns true or false on whether the job is ready to run now."
+  [job]
+  {:post [(or (true? %) (false? %))]}
+  (let [{:keys [status cache-expires-at] :as result} (ccache/get-if-present
+                                                       job-invocations-cache
+                                                       :job/uuid
+                                                       job)
+        expired? (and cache-expires-at (t/after? (t/now) cache-expires-at))]
+    ; Fast path, if it cached, has an expiration (i.e., it was found), and its not
+    ; expired, and it is accepted, then we're done.
+    (if (and result (not expired?))
+      (= status :accepted)
+      (filter-job-invocations-miss job))))
+
+(defn hook-jobs-submission
+  [jobs]
+  "Run the hooks for a set of jobs at submission time."
+  (let [deadline (->> submission-hook-batch-timeout-seconds
+                      ; One submission can include multiple jobs that must all be checked.
+                      ; Self-imposed deadline to get them all checked.
+                      (t/plus- (t/now)))
+        do-one-job (fn do-one-job [job-map]
+                     (let [now (t/now)
+                           status (if (t/before? now deadline)
+                                    (check-job-submission hook-object job-map)
+                                    ; Running out of time, do the default.
+                                    (check-job-submission-default hook-object))]
+                       (or status default-accept)))
+        results (apply list (map do-one-job jobs))
```

Comment:
What is the `apply list` for?

Reply:
I guess I should have been more explicit, why do we need a list instead of a seq?